### PR TITLE
Assembler synchronizer integration

### PR DESCRIPTION
**Pull request by DorKatzelnick**
#783

### CODE REVIEW

File: node/assembler/assembler.go [hunk 1]
```diff
@@ -226,43 +235,69 @@ func (a *Assembler) initFromConfig(
...
+			CommonBlock: configBlock,
+			DecisionNum: 0,
+			BatchIndex:  0,
+			BatchCount:  1,
+		}
+		a.ledger.AppendConfig(ordInfo)
+		a.logger.Infof("Appended genesis block, header digest: %s", hex.EncodeToString(protoutil.BlockHeaderHash(configBlock.GetHeader())))
+	} else if configBlockNumber > ledgerHeight {
+		// genesis block is block number 0, so if config block number is higher than ledger height, it means the assembler needs to sync to the config block before starting.
+		a.logger.Warnf("Config block number %d is higher than current ledger height %d, assembler will need to sync to the config block", configBlockNumber, ledgerHeight)
```
Comment (tock-ibm) [inline node/assembler/assembler.go]: Info, not  warning

File: node/assembler/assembler.go [hunk 2]
```diff
@@ -226,43 +235,69 @@ func (a *Assembler) initFromConfig(
...
+			targetHeight,
+			configBlock,
+		)
+
+		err := synchronizer.Sync()
+		if err != nil {
+			a.logger.Panicf("error while syncing to config block %v", err)
+		}
+	} else {
+		// todo - should panic?
```
Comment (tock-ibm) [inline node/assembler/assembler.go]: Not panic - this is the normal case when restarting after boot or join

File: node/assembler/assembler_test.go
Comment (tock-ibm) [inline node/assembler/assembler_test.go:1]: unit test

File: node/assembler/synchronizer/synchronizer_bft.go [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+		return nil, errors.Wrapf(err, "failed to create BlockVerificationAssistant")
+	}
+
+	clientConfig := a.ClusterDialer.Config // The cluster and block puller use slightly different options
+	clientConfig.AsyncConnect = false
+	clientConfig.SecOpts.VerifyCertificate = nil
+
+	block, _ := pem.Decode(clientConfig.SecOpts.Certificate)
+	tlsCertHash := util.ComputeSHA256(block.Bytes)
+
```
[the remark this answers is not in the document]
Comment (DorKatzelnick) [reply node/assembler/synchronizer/synchronizer_bft.go:245]: fixed

File: node/assembler/synchronizer/synchronizer_factory.go [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+			SelfPartyID:         selfID,
+			TargetHeight:        targetHeight,
+			Support:             support,
+			CryptoProvider:      bccsp,
+			ClusterDialer:       &comm.PredicateDialer{Config: support.ClientConfig()},
+			LocalConfigCluster:  localConfigCluster,
+			BlockPullerFactory:  &GenesisFetcherCreator{},
+			VerifierFactory:     &noopVerifierCreator{}, // TODO replace with real verifer
+			BFTDelivererFactory: &bftDelivererCreator{},
+			Logger:              logger,
```
[the remark this answers is not in the document]
Comment (DorKatzelnick) [reply node/assembler/synchronizer/synchronizer_factory.go:95]: future PR

File: node/assembler/assembler.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (DorKatzelnick) [reply node/assembler/assembler.go]: fixed

File: node/assembler/assembler_test.go [hunk 5]
```diff
@@ -278,21 +283,9 @@ func TestAssembler_SkipNonGenesisConfigBlock(t *testing.T) {
...
-	require.NoError(t, err)
-	require.Equal(t, uint64(0), al.Ledger.Height())
-	genesisBlock, err := al.Ledger.RetrieveBlockByNumber(blockNumber)
-	require.ErrorContains(t, err, "no such block number")
-	require.Nil(t, genesisBlock)
-	al.Close()
+	// Act & Assert: starting with a non-genesis config block and an empty ledger should panic
+	// because the assembler cannot operate without first syncing the missing blocks.
+	require.Panics(t, func() { test.StartAssembler() })
 }
```
[the remark this answers is not in the document]
Comment (DorKatzelnick) [reply node/assembler/assembler_test.go]: fixed

File: node/assembler/assembler_test.go
Comment (DorKatzelnick) [reply node/assembler/assembler_test.go:1]: added unit tests

File: node/assembler/synchronizer/synchronizer_bft.go [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+	if startHeight == 0 {
+		genesisBlock, err := a.fetchGenesisBlock()
+		if err != nil {
+			a.Logger.Panicf("Cannot join the cluster: %s", errors.Wrap(err, "failed to fetch genesis block"))
+		}
+		a.Support.WriteConfigBlock(genesisBlock)
+		startHeight = a.Support.Height()
+		a.Logger.Infof("Fetched and wrote genesis block, new height: %d, party: %d", startHeight, a.SelfPartyID)
+	}
```
[the remark this answers is not in the document]
Comment (DorKatzelnick) [reply node/assembler/synchronizer/synchronizer_bft.go:130]: next PR will add unit tests that checks the real synchronizer
here, added unit tests for the initLedger functionality

File: node/assembler/assembler_test.go [hunk 7]
```diff
@@ -278,21 +286,10 @@ func TestAssembler_SkipNonGenesisConfigBlock(t *testing.T) {
...
-	al, err := node_ledger.NewAssemblerLedger(test.logger, test.ledgerDir)
-	require.NoError(t, err)
-	require.Equal(t, uint64(0), al.Ledger.Height())
-	genesisBlock, err := al.Ledger.RetrieveBlockByNumber(blockNumber)
-	require.ErrorContains(t, err, "no such block number")
-	require.Nil(t, genesisBlock)
-	al.Close()
+	// Act & Assert: starting with a non-genesis config block and an empty ledger should panic
+	// because the assembler cannot operate without first syncing the missing blocks.
+	require.Panics(t, func() { test.StartAssembler() })
```
Comment (tock-ibm) [inline node/assembler/assembler_test.go]: Since you did not specify what the factory returns, this probably panics because the empty mock  synchronizer ends ok, but does not fetch any block?
Is that what you want to check?
Can you require panic with value to verify where exactly the panic comes from?

Comment (tock-ibm) [reply node/assembler/assembler_test.go]: probably "Assembler ledger is empty after initialization, this should not happen"

File: node/assembler/assembler.go [hunk 8]
```diff
@@ -226,43 +235,66 @@ func (a *Assembler) initFromConfig(
...
+			&AssemblerSupportAdapter{assembler: a},
+			factory.GetDefault(),
+			targetHeight,
+			configBlock,
+		)
+
+		err := synchronizer.Sync()
+		if err != nil {
+			a.logger.Panicf("error while syncing to config block %v", err)
+		}
```
[the remark this answers is not in the document]
Comment (tock-ibm) [reply node/assembler/assembler.go]: Good point. @DorKatzelnick

File: node/assembler/assembler_test.go (hunk 7 above)
Comment (DorKatzelnick) [reply node/assembler/assembler_test.go]: no need for this test as it a duplicate of TestAssembler_InitLedgerPanicsWhenLedgerEmptyAfterSync

File: node/assembler/assembler.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (DorKatzelnick) [reply node/assembler/assembler.go]: fixed

**Review by tock-ibm — APPROVED** [review]
